File: 3812-smallest-palindromic-rearrangement-i/smallest-palindromic-rearrangement-i.cpp
class Solution {
public:
    string smallestPalindrome(string s) {
        vector<int>freq(26, 0);

        for(auto ch: s) freq[ch - 'a'] += 1;

        string middle;
        string left;

        for(int i = 0; i < 26; i++){
            left += string(freq[i]/2, char(i + 'a'));

            if(freq[i] % 2 != 0) middle = char('a' + i);
        }

        string right = left;
        reverse(right.begin(), right.end());

        return left + middle + right;
    }
};

File: 3812-smallest-palindromic-rearrangement-i/smallest-palindromic-rearrangement-i.py
class Solution(object):
    def smallestPalindrome(self, s):
        """
        :type s: str
        :rtype: str
        """
        freq = [0] * 26
        for ch in s:
            freq[ord(ch) - ord('a')] += 1
        
        left = []
        middle = ""

        for i in range(26):
            left.append(chr(i + ord('a')) * (freq[i] // 2))

            if freq[i] % 2 != 0:
                middle = chr(i + ord('a'))
        left = "".join(left)
        right = left[::-1]

        return left + middle + right
        